src/analysis/normed_space/star/multiplier.lean

```diff
@@ -0,0 +1,478 @@
+@[simp] lemma nat_cast_to_prod (n : ℕ) : (n : 𝓜(𝕜 , A)).to_prod = n := rfl
+@[simp] lemma int_cast_to_prod (n : ℤ) : (n : 𝓜(𝕜 , A)).to_prod = n := rfl
+@[simp] lemma pow_to_prod (n : ℕ) (a : 𝓜(𝕜, A)) : (a ^ n).to_prod = a.to_prod ^ n := rfl
+
+lemma add_fst (a b : 𝓜(𝕜, A)) : (a + b).fst = a.fst + b.fst := rfl
+lemma add_snd (a b : 𝓜(𝕜, A)) : (a + b).snd = a.snd + b.snd := rfl
+lemma zero_fst : (0 : 𝓜(𝕜, A)).fst = 0 := rfl
+lemma zero_snd : (0 : 𝓜(𝕜, A)).snd = 0 := rfl
+lemma neg_fst (a : 𝓜(𝕜, A)) : (-a).fst = -a.fst := rfl
+lemma neg_snd (a : 𝓜(𝕜, A)) : (-a).snd = -a.snd := rfl
+lemma sub_fst (a b : 𝓜(𝕜, A)) : (a - b).fst = a.fst - b.fst := rfl
+lemma sub_snd (a b : 𝓜(𝕜, A)) : (a - b).snd = a.snd - b.snd := rfl
+lemma one_fst : (1 : 𝓜(𝕜, A)).fst = 1 := rfl
+lemma one_snd : (1 : 𝓜(𝕜, A)).snd = 1 := rfl
+@[simp] lemma mul_fst (a b : 𝓜(𝕜, A)) : (a * b).fst = a.fst * b.fst := rfl
+@[simp] lemma mul_snd (a b : 𝓜(𝕜, A)) : (a * b).snd = b.snd * a.snd := rfl
+lemma nat_cast_fst (n : ℕ) : (n : 𝓜(𝕜 , A)).fst = n := rfl
+lemma nat_cast_snd (n : ℕ) : (n : 𝓜(𝕜 , A)).snd = n := rfl
+lemma int_cast_fst (n : ℤ) : (n : 𝓜(𝕜 , A)).fst = n := rfl
+lemma int_cast_snd (n : ℤ) : (n : 𝓜(𝕜 , A)).snd = n := rfl
+lemma pow_fst (n : ℕ) (a : 𝓜(𝕜, A)) : (a ^ n).fst = a.fst ^ n := rfl
+lemma pow_snd (n : ℕ) (a : 𝓜(𝕜, A)) : (a ^ n).snd = a.snd ^ n := rfl
+
+/-- The natural injection from `double_centralizer.to_prod` except the second coordinate inherits
+`mul_opposite.op`. The ring structure on `𝓜(𝕜, A)` is the pullback under this map. -/
+def to_prod_mul_opposite : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ :=
+λ a, (a.fst, mul_opposite.op a.snd)
+
+lemma to_prod_mul_opposite_injective :
+  function.injective (to_prod_mul_opposite : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ) :=
+λ a b h, let h' := prod.ext_iff.mp h in ext _ _ $ prod.ext h'.1 $ mul_opposite.op_injective h'.2
+
+/-- The ring structure is inherited as the pullback under the injective map
+`double_centralizer.to_prod_mop : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ` -/
+instance : ring 𝓜(𝕜, A) :=
+to_prod_mul_opposite_injective.ring _
+  rfl rfl (λ _ _, rfl) (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl)
+  (λ x n, prod.ext rfl $ mul_opposite.op_smul _ _)
+  (λ x n, prod.ext rfl $ mul_opposite.op_smul _ _)
+  (λ x n, prod.ext rfl $ mul_opposite.op_pow _ _)
+  (λ _, rfl) (λ _, rfl)
+
+/-- The canonical map `double_centralizer.to_prod` as an additive group homomorphism. -/
+def to_prod_hom : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A) :=
+{ to_fun := to_prod,
+  map_zero' := rfl,
+  map_add' := λ x y, rfl }
+
+/-- The module structure is inherited as the pullback under the additive group monomorphism
+`double_centralizer.to_prod : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A)` -/
+instance {S : Type*} [semiring S] [module S A] [smul_comm_class 𝕜 S A]
+  [has_continuous_const_smul S A] [is_scalar_tower S A A] [smul_comm_class S A A] :
+  module S 𝓜(𝕜, A) :=
+function.injective.module S to_prod_hom ext (λ x y, rfl)
+
+instance : algebra 𝕜 𝓜(𝕜, A) :=
```

Comment:
How much work is generalizing to

```suggestion
instance : algebra S 𝓜(𝕜, A) :=
```

Reply:
I thought about this when generalizing the module instanced, but I *think* this is actually hard as is because [continuous_linear_map.algebra](https://leanprover-community.github.io/mathlib_docs/topology/algebra/module/basic.html#continuous_linear_map.algebra) isn't general enough to handle it.

Reply:
Can you leave a todo then?